File: salesflow-app/src/backend/app/api/goals.py
"""
╔════════════════════════════════════════════════════════════════════════════╗
║  GOALS API ROUTER                                                          ║
║  FastAPI Endpoints für Goal-Berechnung                                     ║
╚════════════════════════════════════════════════════════════════════════════╝

Endpoints:
- POST /goals/calculate - Berechnet Goal-Breakdown
- POST /goals/daily-targets - Berechnet tägliche Targets
- GET  /goals/verticals - Liste verfügbarer Verticals
- GET  /goals/kpis/{vertical_id} - KPIs für ein Vertical
"""

from typing import Any, Optional
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field

from ..domain.goals.types import VerticalId, GoalKind
from ..domain.goals.service import get_goal_service, GoalCalculationService


# ═══════════════════════════════════════════════════════════════════════════
# REQUEST / RESPONSE MODELS
# ═══════════════════════════════════════════════════════════════════════════

class GoalCalculateRequest(BaseModel):
    """Request für Goal-Berechnung."""
    vertical_id: str = Field(..., description="Vertical ID (network_marketing, real_estate, etc.)")
    goal_kind: str = Field(..., description="Art des Ziels (income, rank, deals, etc.)")
    target_value: float = Field(..., gt=0, description="Zielwert")
    timeframe_months: int = Field(..., ge=1, le=60, description="Zeitraum in Monaten")
    current_value: float = Field(default=0, ge=0, description="Aktueller Wert")
    vertical_meta: dict[str, Any] = Field(default_factory=dict, description="Vertical-spezifische Daten")
    
    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "vertical_id": "network_marketing",
                    "goal_kind": "income",
                    "target_value": 2000,
                    "timeframe_months": 6,
                    "vertical_meta": {"comp_plan_id": "zinzino"}
                },
                {
                    "vertical_id": "real_estate",
                    "goal_kind": "income",
                    "target_value": 5000,
                    "timeframe_months": 12,
                    "vertical_meta": {"avg_commission": 8000}
                }
            ]
        }
    }


class GoalBreakdownResponse(BaseModel):
    """Response für Goal-Breakdown."""
    vertical_id: str
    goal_kind: str
    timeframe_months: int
    primary_units: float
    secondary_units: float
    required_volume: float
    per_month_volume: float
    per_week_volume: float
    per_day_volume: float
    vertical_details: dict[str, Any]
    notes: str


class DailyTargetsRequest(BaseModel):
    """Request für Daily Targets Berechnung."""
    breakdown: GoalBreakdownResponse
    config: Optional[dict] = None


class DailyTargetsResponse(BaseModel):
    """Response für Daily/Weekly Targets."""
    daily: dict[str, int]
    weekly: dict[str, Any]


class VerticalInfoResponse(BaseModel):
    """Info über ein Vertical."""
    id: str
    label: str


class KpiDefinitionResponse(BaseModel):
    """KPI Definition."""
    id: str
    label: str
    description: str
    unit: str
    icon: str
    color: str


# ═══════════════════════════════════════════════════════════════════════════
# ROUTER
# ═══════════════════════════════════════════════════════════════════════════

router = APIRouter(prefix="/goals", tags=["Goals"])


def get_service() -> GoalCalculationService:
    """Dependency für GoalCalculationService."""
    return get_goal_service()


@router.post("/calculate", response_model=GoalBreakdownResponse)
async def calculate_goal(
    request: GoalCalculateRequest,
    service: GoalCalculationService = Depends(get_service),
):
    """
    Berechnet den Goal-Breakdown.
    
    Nimmt ein Ziel (Income, Rank, Deals, etc.) und berechnet:
    - Benötigte Einheiten (Kunden, Partner, Deals, etc.)
    - Volumen pro Tag/Woche/Monat
    - Vertical-spezifische Details
    """
    try:
        # Konvertiere zu internem GoalInput
        from ..domain.goals.types import GoalInput, VerticalId, GoalKind
        
        goal_input = GoalInput(
            vertical_id=VerticalId(request.vertical_id),
            goal_kind=GoalKind(request.goal_kind),
            target_value=request.target_value,
            timeframe_months=request.timeframe_months,
            current_value=request.current_value,
            vertical_meta=request.vertical_meta,
        )
        
        breakdown = service.calculate_goal(goal_input)
        
        return GoalBreakdownResponse(
            vertical_id=breakdown.vertical_id.value if hasattr(breakdown.vertical_id, 'value') else breakdown.vertical_id,
            goal_kind=breakdown.goal_kind.value if hasattr(breakdown.goal_kind, 'value') else breakdown.goal_kind,
            timeframe_months=breakdown.timeframe_months,
            primary_units=breakdown.primary_units,
            secondary_units=breakdown.secondary_units,
            required_volume=breakdown.required_volume,
            per_month_volume=breakdown.per_month_volume,
            per_week_volume=breakdown.per_week_volume,
            per_day_volume=breakdown.per_day_volume,
            vertical_details=breakdown.vertical_details,
            notes=breakdown.notes,
        )
        
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Berechnung fehlgeschlagen: {str(e)}")


@router.post("/daily-targets", response_model=DailyTargetsResponse)
async def calculate_daily_targets(
    request: DailyTargetsRequest,
    service: GoalCalculationService = Depends(get_service),
):
    """
    Berechnet tägliche und wöchentliche Aktivitäts-Targets.
    
    Basierend auf einem Goal-Breakdown werden konkrete Targets berechnet:
    - Neue Kontakte pro Tag
    - Follow-ups pro Tag
    - Reaktivierungen pro Tag
    - Wöchentliche Ziele
    """
    try:
        from ..domain.goals.types import GoalBreakdown, VerticalId, GoalKind, DailyFlowConfig
        
        breakdown = GoalBreakdown(
            vertical_id=VerticalId(request.breakdown.vertical_id),
            goal_kind=GoalKind(request.breakdown.goal_kind),
            timeframe_months=request.breakdown.timeframe_months,
            primary_units=request.breakdown.primary_units,
            secondary_units=request.breakdown.secondary_units,
            required_volume=request.breakdown.required_volume,
            per_month_volume=request.breakdown.per_month_volume,
            per_week_volume=request.breakdown.per_week_volume,
            per_day_volume=request.breakdown.per_day_volume,
            vertical_details=request.breakdown.vertical_details,
            notes=request.breakdown.notes,
        )
        
        config = None
        if request.config:
            config = DailyFlowConfig(**request.config)
        
        targets = service.calculate_daily_targets(breakdown, config)
        
        return DailyTargetsResponse(
            daily=targets["daily"],
            weekly=targets["weekly"],
        )
        
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Berechnung fehlgeschlagen: {str(e)}")


@router.get("/verticals", response_model=list[VerticalInfoResponse])
async def list_verticals(
    service: GoalCalculationService = Depends(get_service),
):
    """
    Liste aller verfügbaren Verticals.
    
    Gibt die IDs und Labels aller registrierten Vertical-Adapter zurück.
    """
    verticals = []
    for vertical_id in service.get_available_verticals():
        adapter = service.get_adapter(vertical_id)
        if adapter:
            verticals.append(VerticalInfoResponse(
                id=vertical_id,
                label=adapter.get_label(),
            ))
    return verticals


@router.get("/kpis/{vertical_id}", response_model=list[KpiDefinitionResponse])
async def get_kpis(
    vertical_id: str,
    service: GoalCalculationService = Depends(get_service),
):
    """
    KPI-Definitionen für ein Vertical.
    
    Gibt die relevanten KPIs für die Dashboard-Anzeige zurück.
    """
    kpis = service.get_kpi_definitions(vertical_id)
    
    if not kpis:
        raise HTTPException(status_code=404, detail=f"Vertical '{vertical_id}' nicht gefunden")
    
    return [
        KpiDefinitionResponse(
            id=kpi.id,
            label=kpi.label,
            description=kpi.description,
            unit=kpi.unit,
            icon=kpi.icon,
            color=kpi.color,
        )
        for kpi in kpis
    ]


@router.get("/conversion-config/{vertical_id}")
async def get_conversion_config(
    vertical_id: str,
    service: GoalCalculationService = Depends(get_service),
):
    """
    Default Conversion-Config für ein Vertical.
    
    Gibt die Standard-Konversionsraten für Daily Flow Berechnungen zurück.
    """
    config = service.get_conversion_config(vertical_id)
    
    if not config:
        raise HTTPException(status_code=404, detail=f"Vertical '{vertical_id}' nicht gefunden")
    
    return {
        "working_days_per_week": config.working_days_per_week,
        "contact_to_primary_unit": config.contact_to_primary_unit,
        "contact_to_secondary_unit": config.contact_to_secondary_unit,
        "followups_per_primary": config.followups_per_primary,
        "followups_per_secondary": config.followups_per_secondary,
        "reactivation_share": config.reactivation_share,
    }

File: salesflow-app/src/types/goals.ts
/**
 * ╔════════════════════════════════════════════════════════════════════════════╗
 * ║  SALES FLOW AI - GOAL TYPES (Vertriebsagnostisch)                         ║
 * ║  OS für jeden Vertrieb: MLM, Immobilien, Finance, Coaching, etc.          ║
 * ╚════════════════════════════════════════════════════════════════════════════╝
 * 
 * Diese Types sind generisch und funktionieren für alle Verticals.
 * Sie ergänzen die bestehenden compensation.ts Types um mehr Flexibilität.
 */

import { z } from 'zod';

// ═══════════════════════════════════════════════════════════════════════════
// VERTICAL ID (erweitert für mehr Branchen)
// ═══════════════════════════════════════════════════════════════════════════

export const VerticalIdSchema = z.enum([
  'network_marketing',
  'real_estate',
  'finance',
  'coaching',
  'b2b_saas',
  'insurance',
  'generic',
]);
export type VerticalId = z.infer<typeof VerticalIdSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// GOAL KIND (erweitert: nicht nur income/rank)
// ═══════════════════════════════════════════════════════════════════════════

export const GoalKindSchema = z.enum([
  'income',   // Ziel-Einkommen pro Monat
  'rank',     // Rang / Status / Level
  'deals',    // Anzahl Abschlüsse / Verkäufe
  'volume',   // Umsatz / Volumen (Credits, Provision, etc.)
  'clients',  // Kundenanzahl
]);
export type GoalKind = z.infer<typeof GoalKindSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// GOAL INPUT
// ═══════════════════════════════════════════════════════════════════════════

export const GoalInputSchema = z.object({
  vertical_id: VerticalIdSchema,
  goal_kind: GoalKindSchema,
  target_value: z.number().positive().describe('Zielwert (€, Anzahl, Volumen)'),
  timeframe_months: z.number().int().min(1).max(60).describe('Zeitraum in Monaten'),
  vertical_meta: z.record(z.any()).default({}).describe(
    'Vertical-spezifische Zusatzinfos (z.B. comp_plan_id für MLM)'
  ),
});
export type GoalInput = z.infer<typeof GoalInputSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// GOAL BREAKDOWN (Generisch statt MLM-spezifisch)
// ═══════════════════════════════════════════════════════════════════════════

export const GoalBreakdownSchema = z.object({
  vertical_id: VerticalIdSchema,
  goal_kind: GoalKindSchema,
  timeframe_months: z.number().int().positive(),
  
  // Generische Einheiten
  primary_units: z.number().describe(
    'Haupteinheiten: MLM=Kunden, Immobilien=Deals, Coaching=Programme'
  ),
  secondary_units: z.number().nullable().optional().describe(
    'Sekundäreinheiten: MLM=Partner, Immobilien=Besichtigungen, etc.'
  ),
  
  // Volumen-Breakdown
  required_volume: z.number().nullable().optional().describe(
    'Gesamt-Volumen (Credits/Umsatz/Prämien)'
  ),
  per_month_volume: z.number(),
  per_week_volume: z.number(),
  per_day_volume: z.number(),
  
  // Vertical-spezifische Details
  vertical_details: z.record(z.any()).default({}).describe(
    'Vertical-spezifisch: Rang-Infos, Plan-ID, etc.'
  ),
  
  notes: z.string().nullable().optional().describe('Erklärung/Debug-Info'),
});
export type GoalBreakdown = z.infer<typeof GoalBreakdownSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// DAILY FLOW CONFIG (Generisch mit primary/secondary statt customer/partner)
// ═══════════════════════════════════════════════════════════════════════════

export const GenericDailyFlowConfigSchema = z.object({
  working_days_per_week: z.number().int().min(1).max(7).default(5),
  
  // Konversionsraten
  contact_to_primary_unit: z.number().min(0).max(1).describe(
    'Kontakt → Kunde/Deal Konversionsrate'
  ),
  contact_to_secondary_unit: z.number().min(0).max(1).nullable().optional().describe(
    'Kontakt → Partner/Besichtigung Rate'
  ),
  
  // Follow-up Annahmen
  followups_per_primary: z.number().int().default(3).describe('Follow-ups bis Abschluss'),
  followups_per_secondary: z.number().int().nullable().optional().default(5),
  
  // Reaktivierung
  reactivation_share: z.number().min(0).max(1).default(0.2).describe(
    'Anteil Reaktivierungen am Gesamt-Flow'
  ),
});
export type GenericDailyFlowConfig = z.infer<typeof GenericDailyFlowConfigSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// DAILY FLOW TARGETS (Generisch)
// ═══════════════════════════════════════════════════════════════════════════

export const GenericDailyFlowTargetsSchema = z.object({
  user_id: z.string().uuid(),
  company_id: z.string(),
  vertical_id: VerticalIdSchema,
  
  // Generische Targets (Labels werden vom Vertical bestimmt)
  new_contacts: z.number().int().describe('Neue Kontakte/Leads pro Tag'),
  followups: z.number().int().describe('Follow-ups pro Tag'),
  reactivations: z.number().int().describe('Reaktivierungen pro Tag'),
  
  // Optional: Vertical-spezifische Targets
  custom_targets: z.record(z.number().int()).default({}).describe(
    "z.B. {'viewings': 2, 'offers': 1} für Makler"
  ),
});
export type GenericDailyFlowTargets = z.infer<typeof GenericDailyFlowTargetsSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// KPI DEFINITION (Erweitert für dynamische UI)
// ═══════════════════════════════════════════════════════════════════════════

export const ExtendedKpiDefinitionSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string().nullable().optional(),
  unit: z.enum(['per_day', 'per_week', 'per_month', 'total', 'currency', 'percent']).default('per_day'),
  icon: z.string().nullable().optional().describe('Lucide icon name'),
  color: z.string().nullable().optional().describe('Tailwind color'),
});
export type ExtendedKpiDefinition = z.infer<typeof ExtendedKpiDefinitionSchema>;

// ═══════════════════════════════════════════════════════════════════════════
// VERTICAL-SPECIFIC UNIT LABELS
// ═══════════════════════════════════════════════════════════════════════════

export interface VerticalUnitLabels {
  primary_unit: string;      // "Kunden" | "Deals" | "Abschlüsse" | "Programme"
  secondary_unit: string;    // "Partner" | "Besichtigungen" | "Leads"
  volume_unit: string;       // "Credits" | "€ Provision" | "Punkte"
}

export const VERTICAL_UNIT_LABELS: Record<VerticalId, VerticalUnitLabels> = {
  network_marketing: {
    primary_unit: 'Kunden',
    secondary_unit: 'Partner',
    volume_unit: 'Credits',
  },
  real_estate: {
    primary_unit: 'Abschlüsse',
    secondary_unit: 'Besichtigungen',
    volume_unit: '€ Provision',
  },
  finance: {
    primary_unit: 'Verträge',
    secondary_unit: 'Beratungsgespräche',
    volume_unit: '€ Provision',
  },
  coaching: {
    primary_unit: 'Klienten',
    secondary_unit: 'Discovery Calls',
    volume_unit: '€ Umsatz',
  },
  b2b_saas: {
    primary_unit: 'Deals',
    secondary_unit: 'Demos',
    volume_unit: '€ ARR',
  },
  insurance: {
    primary_unit: 'Policen',
    secondary_unit: 'Beratungen',
    volume_unit: '€ Provision',
  },
  generic: {
    primary_unit: 'Abschlüsse',
    secondary_unit: 'Kontakte',
    volume_unit: '€',
  },
};

// ═══════════════════════════════════════════════════════════════════════════
// DEFAULT CONVERSION RATES PER VERTICAL
// ═══════════════════════════════════════════════════════════════════════════

export const DEFAULT_CONVERSION_RATES: Record<VerticalId, GenericDailyFlowConfig> = {
  network_marketing: {
    working_days_per_week: 5,
    contact_to_primary_unit: 0.20,    // 20% werden Kunde
    contact_to_secondary_unit: 0.05,  // 5% werden Partner
    followups_per_primary: 3,
    followups_per_secondary: 5,
    reactivation_share: 0.20,
  },
  real_estate: {
    working_days_per_week: 6,
    contact_to_primary_unit: 0.03,    // 3% Deal-Abschluss
    contact_to_secondary_unit: 0.25,  // 25% kommen zur Besichtigung
    followups_per_primary: 8,
    followups_per_secondary: 3,
    reactivation_share: 0.15,
  },
  finance: {
    working_days_per_week: 5,
    contact_to_primary_unit: 0.15,    // 15% Vertragsabschluss
    contact_to_secondary_unit: 0.40,  // 40% zum Beratungsgespräch
    followups_per_primary: 4,
    followups_per_secondary: 2,
    reactivation_share: 0.25,
  },
  coaching: {
    working_days_per_week: 5,
    contact_to_primary_unit: 0.10,    // 10% werden Klient
    contact_to_secondary_unit: 0.30,  // 30% zum Discovery Call
    followups_per_primary: 5,
    followups_per_secondary: 2,
    reactivation_share: 0.20,
  },
  b2b_saas: {
    working_days_per_week: 5,
    contact_to_primary_unit: 0.05,    // 5% Deal Close
    contact_to_secondary_unit: 0.15,  // 15% zur Demo
    followups_per_primary: 6,
    followups_per_secondary: 3,
    reactivation_share: 0.10,
  },
  insurance: {
    working_days_per_week: 5,
    contact_to_primary_unit: 0.12,    // 12% Police
    contact_to_secondary_unit: 0.35,  // 35% Beratung
    followups_per_primary: 4,
    followups_per_secondary: 2,
    reactivation_share: 0.30,
  },
  generic: {
    working_days_per_week: 5,
    contact_to_primary_unit: 0.10,
    contact_to_secondary_unit: 0.20,
    followups_per_primary: 4,
    followups_per_secondary: 3,
    reactivation_share: 0.20,
  },
};

// ═══════════════════════════════════════════════════════════════════════════
// TYPE GUARDS & VALIDATORS
// ═══════════════════════════════════════════════════════════════════════════

export function isValidVerticalId(value: string): value is VerticalId {
  return VerticalIdSchema.safeParse(value).success;
}

export function isValidGoalKind(value: string): value is GoalKind {
  return GoalKindSchema.safeParse(value).success;
}

export function validateGoalInput(input: unknown): GoalInput {
  return GoalInputSchema.parse(input);
}

export function validateGoalBreakdown(breakdown: unknown): GoalBreakdown {
  return GoalBreakdownSchema.parse(breakdown);
}

// ═══════════════════════════════════════════════════════════════════════════
// HELPER FUNCTIONS
// ═══════════════════════════════════════════════════════════════════════════

/**
 * Holt die passenden Unit-Labels für ein Vertical
 */
export function getUnitLabels(verticalId: VerticalId): VerticalUnitLabels {
  return VERTICAL_UNIT_LABELS[verticalId] ?? VERTICAL_UNIT_LABELS.generic;
}

/**
 * Holt die Default-Konversionsraten für ein Vertical
 */
export function getDefaultConversionRates(verticalId: VerticalId): GenericDailyFlowConfig {
  return DEFAULT_CONVERSION_RATES[verticalId] ?? DEFAULT_CONVERSION_RATES.generic;
}

/**
 * Berechnet tägliche Targets aus einem GoalBreakdown
 */
export function calculateDailyTargetsFromBreakdown(
  breakdown: GoalBreakdown,
  config: GenericDailyFlowConfig
): GenericDailyFlowTargets {
  const workingDaysPerMonth = config.working_days_per_week * 4.33;
  
  // Wie viele primary units pro Tag?
  const primaryUnitsPerDay = breakdown.primary_units / breakdown.timeframe_months / workingDaysPerMonth;
  
  // Kontakte = primary units / conversion rate
  const newContactsPerDay = Math.ceil(primaryUnitsPerDay / config.contact_to_primary_unit);
  
  // Follow-ups basierend auf Pipeline
  const followupsPerDay = Math.ceil(primaryUnitsPerDay * config.followups_per_primary);
  
  // Reaktivierungen
  const reactivationsPerDay = Math.ceil(newContactsPerDay * config.reactivation_share);
  
  return {
    user_id: '', // Wird beim Speichern gesetzt
    company_id: '', // Wird beim Speichern gesetzt
    vertical_id: breakdown.vertical_id,
    new_contacts: newContactsPerDay,
    followups: followupsPerDay,
    reactivations: reactivationsPerDay,
    custom_targets: {},
  };
}

